zerver/views/messages.py

```diff
@@ -1582,11 +1583,63 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
 
         mention_user_ids = message.mentions_user_ids
 
-    number_changed = do_update_message(user_profile, message, topic_name,
-                                       propagate_mode, content, rendered_content,
+    new_stream = None
+    old_stream = None
+    number_changed = 0
+
+    if new_stream_id is not None:
+        if not user_profile.is_realm_admin:
+            raise JsonableError(_("You don't have permission to move this message"))
+        if content is not None:
+            raise JsonableError(_("Cannot change message content while changing stream"))
+        if topic_name is not None:
+            # TODO: Add support for changing the topic together with
+            # the stream; it seems useful and should easily fit within
+            # an extended topic-edit UI.
+            raise JsonableError(_("Cannot change topic while changing stream"))
+
+        old_topic_name = message.topic_name()
+
+        old_stream = get_stream_by_id(message.recipient.type_id)
+        new_stream = get_stream_by_id(new_stream_id)
+
+        if not (old_stream.is_public() and new_stream.is_public()):
+            # We'll likely decide to relax this condition in the
+            # future; it just requires more care with details like the
+            # breadcrumb messages.
+            raise JsonableError(_("Streams must be public"))
+
+    number_changed = do_update_message(user_profile, message, new_stream,
+                                       topic_name, propagate_mode,
+                                       content, rendered_content,
                                        prior_mention_user_ids,
                                        mention_user_ids, mention_data)
 
+    if (number_changed > 0 and new_stream_id is not None) and (
+        new_stream is not None and old_stream is not None
+    ):
+        # Since moving content between streams is highly disruptive,
+        # it's worth adding a couple tombstone messages showing what
+        # happened.
+        sender = get_system_bot(settings.NOTIFICATION_BOT)
+
+        # Send a notification to the new stream that the topic was moved.
+        user_mention = "**_%s|%s**" % (user_profile.full_name, user_profile.id)
```

Comment:
These links are missing `@s`, which is why the notifications look ugly like this: 
![image](https://user-images.githubusercontent.com/2746074/75937563-19ab6b00-5e3a-11ea-9f1c-ccbf9688a748.png)